python/lsst/daf/butler/core/datasets/ref.py

```diff
@@ -628,6 +628,56 @@ def overrideStorageClass(self, storageClass: str | StorageClass) -> DatasetRef:
             conform=False,
         )
 
+    def is_compatible_with(self, ref: DatasetRef) -> bool:
+        """Determine if the given `DatasetRef` is compatible with this one.
+
+        Parameters
+        ----------
+        other : `DatasetRef`
+            Dataset ref to check.
+
+        Returns
+        -------
+        is_compatible : `bool`
+            Returns `True` if the other dataset ref is either the same as this
+            or the dataset type associated with the other is compatible with
+            this one and the dataId and dataset ID match.
+
+        Notes
+        -----
+        Compatibility requires that the dataId and dataset ID match and the
+        `DatasetType` is compatible. Compatibility is defined as the storage
+        class associated with the dataset type of the other ref can be
+        converted to this storage class.
+
+        Specifically this means that if you have done:
+
+        .. code-block:: py
+
+            new_ref = ref.overrideStorageClass(sc)
+
+        and this is successful, then the guarantee is that:
+
+        .. code-block:: py
+
+            assert ref.is_compatible_with(new_ref) is True
+
+        since we know that the python type associated with the new ref can
+        be converted to the original python type. The reverse is not guaranteed
+        and depends on whether bidirectional converters have been registered.
+        """
+        if not isinstance(ref, DatasetRef):
+            # mypy thinks this is impossible but we need to prevent
+            # an AttributeError if possible.
```

Comment:
Not clear where the AttributeError comes from?

Reply:
It will come from the `ref.id` below if you pass the wrong thing here. mypy never expects you to pass the wrong thing here so is convinced it can't happen.

Reply:
But this is a completely new method, don't we control what is being passed to it? And if someone passes and incorrect type then they probably deserve AttributeError, not? Otherwise if we allow other types maybe just declare argument as `ref: object`?

Reply:
I was being defensive since it felt to me that this was a lot like `__eq__` which does have to worry about any random thing being given to it. If you feel that AttributeError is fine if someone passes in a non-ref I'll change it.

Reply:
I think that checking a non-ref to be compatible with a ref is a meaningless operation and no one is going to ask those questions, so an exception should be fine here. And Python 4 should probably raise TypeError if an actual argument type is different from annotation 🙂